packages/expo-network/android/src/main/java/expo/modules/network/NetworkModule.kt

```diff
@@ -0,0 +1,158 @@
+    ConnectivityManager.TYPE_ETHERNET -> NetworkStateType.ETHERNET
+    ConnectivityManager.TYPE_WIMAX -> NetworkStateType.WIMAX
+    ConnectivityManager.TYPE_VPN -> NetworkStateType.VPN
+    else -> NetworkStateType.UNKNOWN
+  }
+
+  private fun getConnectionType(netCapabilities: NetworkCapabilities?): NetworkStateType =
+    if (netCapabilities == null) NetworkStateType.UNKNOWN
+    else if (netCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR)) NetworkStateType.CELLULAR
+    else if (netCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) || netCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI_AWARE)) NetworkStateType.WIFI
+    else if (netCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_BLUETOOTH)) NetworkStateType.BLUETOOTH
+    else if (netCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET)) NetworkStateType.ETHERNET
+    else if (netCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_VPN)) NetworkStateType.VPN
+    else NetworkStateType.UNKNOWN
+
+  private fun rawIpToString(ip: Int): String {
+    // Convert little-endian to big-endian if needed
+    val ip = if (ByteOrder.nativeOrder() == ByteOrder.LITTLE_ENDIAN) {
+      Integer.reverseBytes(ip)
+    } else {
+      ip
+    }
+
+    var ipByteArray = BigInteger.valueOf(ip.toLong()).toByteArray()
+    if (ipByteArray.size < 4) {
+      ipByteArray = frontPadWithZeros(ipByteArray)
+    }
+
+    return try {
+      InetAddress.getByAddress(ipByteArray).hostAddress
+    } catch (e: UnknownHostException) {
+      "0.0.0.0"
+    }
+  }
+
+  @ExpoMethod
+  fun getNetworkStateAsync(promise: Promise) {
+    val result = Bundle()
+    val connectivityManager = appContext.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
+
+    try {
+      if (Build.VERSION.SDK_INT < 29) { // use getActiveNetworkInfo before api level 29
+        val netInfo = connectivityManager.activeNetworkInfo
+        val connectionType = getConnectionType(netInfo)
+
+        result.apply {
+          putBoolean("isInternetReachable", netInfo!!.isConnected)
+          putString("type", connectionType.value)
+          putBoolean("isConnected", connectionType.isDefined)
+        }
+
+        promise.resolve(result)
+      } else {
+        val network = connectivityManager.activeNetwork
+        val isInternetReachable = network != null
+
+        var connectionType: NetworkStateType? = if (isInternetReachable) {
+          val netCapabilities = connectivityManager.getNetworkCapabilities(network)
+          getConnectionType(netCapabilities)
+        } else {
+          null
+        }.also { connectionType ->
+          result.apply {
+            putString("type", connectionType?.value ?: NetworkStateType.NONE.value)
+            putBoolean("isInternetReachable", isInternetReachable)
+            putBoolean("isConnected", connectionType != null && connectionType.isDefined)
+          }
+        }
```

Comment:
The `also` scope function looks safe to remove:

```suggestion
        val connectionType = if (isInternetReachable) {
          val netCapabilities = connectivityManager.getNetworkCapabilities(network)
          getConnectionType(netCapabilities)
        } else {
          null
        }

        result.apply {
          putString("type", connectionType?.value ?: NetworkStateType.NONE.value)
          putBoolean("isInternetReachable", isInternetReachable)
          putBoolean("isConnected", connectionType != null && connectionType.isDefined)
        }
```